Docs/893 tunable properties page (#963)
redpanda-data/documentation: docs/platform/reference/tunable-properties.mdx

---
title: Tunable Configuration Properties
toc_min_heading_level: 2
toc_max_heading_level: 3
---

<head>
  <meta
    name="title"
    content="Tunable configuration properties | Redpanda Docs"
  />
  <meta name="description" content="Reference of tunable cluster properties." />
</head>

Tunable properties are cluster properties that are specifically designed for tuning a cluster.

> For information on how to edit tunable properties, refer to [Configuring Cluster Properties](../../cluster-administration/cluster-property-configuration).

Reviewer: ```suggestion
For information on how to edit tunable properties, see [Configuring Cluster Properties](../../cluster-administration/cluster-property-configuration).
```

## Cloud storage properties

### cloud_storage_cache_check_interval

Timeout to check if cache eviction should be triggered.

**Units**: milliseconds

**Default**: 30000 (30 seconds)

---

### cloud_storage_enable_compacted_topic_reupload

Enable re-upload of data for compacted topics.

**Type**: boolean

**Default**: true

---

### cloud_storage_enable_remote_read

Enable remote reads for all topics.

**Type**: boolean

**Default**: false

---

### cloud_storage_enable_remote_write

Enable remote writes for all topics.

**Type**: boolean

**Default**: false

---

### cloud_storage_housekeeping_interval_ms

Interval for cloud storage housekeeping tasks.

**Units**: milliseconds

**Default**: 300000 (5 minutes)

---

### cloud_storage_initial_backoff_ms

Initial backoff time for exponential backoff algorithm.

**Units**: milliseconds

**Default**: 100

---

### cloud_storage_manifest_upload_timeout_ms

Manifest upload timeout.

**Units**: milliseconds

**Default**: 10000 (10 seconds)

---

### cloud_storage_max_connection_idle_time_ms

Maximum HTTPS connection idle time.

**Units**: milliseconds

**Default**: 5000

---

### cloud_storage_max_materialized_segments_per_shard

> Maximum number of concurrent readers of remote data per shard (CPU core).  If `null`, the value of [topic_partitions_per_shard](#topic_partitions_per_shard) multiplied by 2 is used.

Reviewer: ```suggestion
Maximum number of concurrent readers of remote data per shard (CPU core).  If `null`, the value of [topic_partitions_per_shard](#topic_partitions_per_shard) multiplied by two is used.
```

**Type**: unsigned integer

**Units**: number of concurrent readers

**Default**: null

---

### cloud_storage_max_readers_per_shard

Maximum concurrent readers of remote data per shard (CPU core).  If `null`, the value of [topic_partitions_per_shard](#topic_partitions_per_shard) is used, with one reader per partition if the shard is at its maximum partition capacity.

**Type**: unsigned integer

**Units**: number of concurrent readers

**Default**: null

---

### cloud_storage_max_segments_pending_deletion_per_partition

> The maximum limit per partition for the number of segments pending deletion from the cloud. Segments can be deleted due to retention or compaction. If this limit is breached and deletion fails, then segments will be orphaned in the cloud and will have to be removed manually.

Reviewer: ```suggestion
The maximum limit per partition for the number of segments pending deletion from the cloud. Segments can be deleted due to retention or compaction. If this limit is breached and deletion fails, then segments are orphaned in the cloud and must be removed manually.
```

**Units**: milliseconds

**Default**: 300000 (5 minutes)

---

### leader_balancer_node_mute_timeout

Timeout used to mute a node, where a muted node is removed from consideration as leadership rebalancing target.

**Units**: milliseconds

**Default**: 20000 (20 seconds)

---

### leader_balancer_transfer_limit_per_shard

Maximum limit per shard (CPU core) on the number of in-progress leadership transfers.

**Type**: unsigned integer

**Units**: number of in-progress leadership transfers per shard

**Default**: 512

**Range**: [1, 2048]

---

### max_concurrent_producer_ids

Maximum number of producer IDs that the resource manager state machine caches in internal state. 

**Type**: unsigned integer

**Units**: number of IDs

**Default**: `std::numeric_limits<uint64_t>::max()`

---

### members_backend_retry_ms

Duration between members backend reconciliation loop retries.

**Units**: milliseconds

**Default**: 5000 

---

### metadata_dissemination_interval_ms

Interval for metadata dissemination batching.

**Units**: milliseconds

**Default**: 3000 

---

### metadata_dissemination_retries

> Number of attempts of looking up a topic's meta data like shard before failing a request.

Reviewer: ```suggestion
Number of attempts looking up a topic's metadata, like shard, before failing a request.
```

---

### rps_limit_move_operations

Request-per-second (rps) rate limit for the cluster controller's move operations.

**Type**: unsigned integer

**Units**: requests per second

**Default**: 1000

---

### rps_limit_node_management_operations

Request-per-second (rps) rate limit for the cluster controller's node management operations.

**Type**: unsigned integer

**Units**: requests per second

**Default**: 1000

--- 

### rps_limit_topic_operations

Request-per-second (rps) rate limit for the cluster controller's topic operations.

**Type**: unsigned integer

**Units**: requests per second

**Default**: 1000

-----------------------------------------------------------------------------------

## Kafka properties

### default_num_windows

Default number of quota tracking windows.

**Units**: number of windows

**Default**: 10

---

### default_window_sec

Default duration of quota tracking window.

**Units**: milliseconds

**Default**: 1000

---

### fetch_reads_debounce_timeout

Duration to wait for the next read in fetch request when minimum number of requested bytes wasn't read.

**Units**: milliseconds

**Default**: 1

---

### fetch_session_eviction_timeout_ms

> Minimum time before an unused session will get evicted. Maximum time after which inactive session will be deleted is twice the given configuration value cache.

Reviewer: ```suggestion
Minimum time before an unused session gets evicted. The maximum time after which an inactive session is deleted is twice the given configuration value cache.
```

Minimum number of I/O and CPU shares that the compaction process can use. 

**Type**: short

**Units**: number of shares

**Default**: 10

---

### compaction_ctrl_p_coeff

Proportional coefficient for compaction PID controller. Must be negative to decrease the compaction backlog when compaction shares increase.

**Type**: double

**Default**: -12.5

**Range**: [..., 0.0)

---

### compaction_ctrl_update_interval_ms

Sampling interval of the compaction PID controller.

**Units**: milliseconds

**Default**: 30000 (30 seconds)

---

### disable_batch_cache

Flag to disable the batch cache in log manager.

**Type**: boolean

**Default**: false

---

### kvstore_flush_interval

Flush interval of the key-value store.

**Units**: milliseconds

**Default**: 10

---

### kvstore_max_segment_size

Maximum size of a segment of the key-value store.

**Type**: unsigned integer

**Units**: bytes

**Default**: 16777216 (16 MiB)

---

### log_segment_size

Default size of log segment. Per topic, it is superseded by topic property `segment.bytes`.

**Type**: unsigned 64-bit integer

**Units**: bytes

**Default**: 1073741824 (1 GiB)

**Range**: [1 MiB, ...]

---

### log_segment_size_max

> Upper limit of topic `segment.bytes`. Higher values will be clamped to this limit. 

Reviewer: ```suggestion
Upper limit of topic `segment.bytes`. Higher values are clamped to this limit. 
```

If `null`, no limit is applied.

**Type**: unsigned 64-bit integer

**Units**: bytes

**Default**: null (no limit applied)

---

### log_segment_size_min

> Lower limit of topic `segment.bytes`. Lower values will be clamped to this limit. 

Reviewer: ```suggestion
Lower limit of topic `segment.bytes`. Lower values are clamped to this limit. 
```

If `null`, no limit is applied.

**Type**: unsigned 64-bit integer

**Units**: bytes

**Default**: null (no limit applied)

---

### log_segment_size_jitter_percent

Configures the amount of size variation added to a segment used for each partition: a percentage value in the range [0, log_segment_size_jitter_percent] is randomly chosen and that percentage of the original segment size limit is added. 

Applying segment size jitter is a good defensive measure to space out the overhead created from segment rolls over time.

**Type**: unsigned short

**Units**: percent of segment size that

**Default**: 5

**Range**: [0, 99]

---

### max_compacted_log_segment_size

Max compacted segment size after consolidation.

**Type**: unsigned integer

**Units**: bytes

**Default**: 5368709120 (5 GiB)

---

### readers_cache_eviction_timeout_ms

> Duration after which inactive readers will be evicted from the cache of readers.

Reviewer: ```suggestion
Duration after which inactive readers are evicted from the cache of readers.
```

**Units**: milliseconds

**Default**: 10000 (10 seconds)

**Range**: >= [reclaim_growth_window](#reclaim_growth_window)

---

### release_cache_on_segment_roll

Flag for whether to release cache when a full segment is rolled.

**Type**: boolean

**Default**: false

---

### replicate_append_timeout_ms

Timeout for append entries RPCs issued while replicating entries.

**Units**: milliseconds

**Default**: 3000

---

### segment_appender_flush_timeout_ms

Maximum duration until buffered data is written.

**Units**: milliseconds

**Default**: 1000

---

### segment_fallocation_step

Size of a file allocation (fallocation) step of a segment.

**Type**: unsigned integer

**Units**: bytes

**Default**: 33554432 (32 MiB)

**Range**: [4096, 1 GiB]. Must be 4096-byte aligned.

---

### storage_compaction_index_memory

Maximum number of bytes that may be used on each shard (CPU core) by compaction index writers.

**Type**: unsigned integer

**Units**: bytes

**Default**: 134217728 (128 MiB)

**Range**: [16 MiB, 100 GiB]

---

### storage_max_concurrent_replay

> Maximum number of partition logs that will be replayed concurrently on startup or flushed concurrently on shutdown.

Reviewer: ```suggestion
Maximum number of partition logs that are replayed concurrently on startup or flushed concurrently on shutdown.
```

**Type**: unsigned integer

**Units**: bytes

**Default**: 8388608 (8 MiB)

--------------------------------------------------------------------------------

## Topic and partition properties

### alter_topic_cfg_timeout_ms

Time to wait for entries replication in controller log when executing alter topic configuration request.

**Units**: milliseconds

**Default**: 5000

 

---

### create_topic_timeout_ms

Timeout to wait for new topic creation.

**Units**: milliseconds

**Default**: 2000

---

### topic_fds_per_partition

The number of file descriptors reserved for topics per partition.

**Type**: integer

**Units**: number of file descriptors

**Default**: 5

**Range**: [1, 1000]

---

### topic_memory_per_partition

The amount of memory reserved for topics per partition.

**Type**: unsigned integer

**Units**: bytes

**Default**: 1048576 (1 MiB)

**Range**: [1, 100 MiB]

---

### topic_partitions_per_shard

Maximum number of partitions which may be allocated per shard (CPU core).

**Type**: unsigned integer

**Units**: number of partitions

**Default**: 7000

**Range**: [16, 131072]

---

### topic_partitions_reserve_shard0

> Reserved partition slots on shard (CPU core) 0 on each node.  If this is >= [topic_partitions_per_shard](#topics_partitions_per_shard), no data partitions will be scheduled on shard 0.

Reviewer: ```suggestion
Reserved partition slots on shard (CPU core) 0 on each node. If this is >= [topic_partitions_per_shard](#topics_partitions_per_shard), no data partitions are scheduled on shard 0.
```

--------------------------------------------------------------------------------

## Wasm properties

:::caution
[Redpanda Data Transforms](../labs/data-transform.mdx), built on top of Wasm, is a lab feature and is not supported for production deployments.
:::

### coproc_max_batch_size

Maximum size of one topic read in coprocessing mode. 

**Units**: bytes

**Default**: 32 KB

---

### coproc_max_inflight_bytes

Maximum size of inflight data when sending to the Wasm engine.

**Type**: unsigned integer 

**Units**: bytes

**Default**: 10485760 (10 MiB)

---

### coproc_max_ingest_bytes

Maximum size of input logs in memory.

**Type**: unsigned integer

**Units**: bytes

**Default**: 655360 (640 KiB)

---

### coproc_offset_flush_interval_ms

Interval for which all coprocessor offsets are flushed to disk.

**Units**: milliseconds

**Default**: 300000 (5 minutes)

----------------------------------------------------------

## Related topics

- [Cluster Configurable Properties](./cluster-properties.mdx)
- [Node Configuration Properties](./node-properties.mdx)
- Benchmark of Redpanda and Apache Kafka [article](https://redpanda.com/blog/fast-and-safe/)

> - Fast distributed transactions with Redpanda [article](https://redpanda.com/blog/fast-transactions/)

Reviewer: We still put links to blogs under a Suggested reading heading, just to differentiate from internal links.

> - Benchmark of Redpanda and Apache Kafka [article](https://redpanda.com/blog/fast-and-safe/)
> - Fast distributed transactions with Redpanda [article](https://redpanda.com/blog/fast-transactions/)

Reviewer: ```suggestion
## Suggested reading
- [Benchmark of Redpanda and Apache Kafka](https://redpanda.com/blog/fast-and-safe/)
- [Fast distributed transactions with Redpanda](https://redpanda.com/blog/fast-transactions/)
```

**Default**: 16

---

### raft_max_recovery_memory

Maximum amount of memory available for raft follower recovery. If `null`, defaults to 10% of total memory.

**Type**: unsigned integer

**Units**: bytes

**Default**: null

**Range**: [32 MiB, ...]

---

### raft_recovery_default_read_size

Default size of a read issued during raft follower recovery.

**Type**: unsigned integer

**Units**: bytes

**Default**: 524288 (512 KiB)

**Range**: [128, 5 MiB]

---

### raft_replicate_batch_window_size

Maximum size of a batch for raft replication.

**Type**: unsigned integer

**Units**: bytes

**Default**: 1048576 (1 MiB)

---

### raft_smp_max_non_local_requests

Maximum number of non-local requests that can execute concurrently on a CPU core (seastar 'shard'). If `null`, defaults to the value determined by the following formula:

```
topic_partitions_per_shard * (266) * (<number of CPU cores> - 1)
```

For details, see [seastar::smp_service_group_config](https://docs.seastar.io/master/namespaceseastar.html#structseastar_1_1smp__service__group__config).

**Type**: unsigned integer

**Units**: number of requests

**Default**: null

---

### raft_timeout_now_timeout_ms

> Timeout of the raft 'timeout now' RPC request.

Reviewer 2: ```suggestion
Timeout of the raft `timeout now` RPC request.
```

Not sure if this should be with backticks; as is looks incorrect. If not backticks, then "timeout now"

**Range**: [0.0, ...]

---

### compaction_ctrl_max_shares

Maximum number of I/O and CPU shares that the compaction process can use. 

**Units**: short

**Units**: number of shares

**Default**: 1000

---

### compaction_ctrl_min_shares

Minimum number of I/O and CPU shares that the compaction process can use. 

**Type**: short

**Units**: number of shares

**Default**: 10

---

### compaction_ctrl_p_coeff

Proportional coefficient for compaction PID controller. Must be negative to decrease the compaction backlog when compaction shares increase.

**Type**: double

**Default**: -12.5

**Range**: [..., 0.0)

---

### compaction_ctrl_update_interval_ms

Sampling interval of the compaction PID controller.

**Units**: milliseconds

**Default**: 30000 (30 seconds)

---

### disable_batch_cache

Flag to disable the batch cache in log manager.

**Type**: boolean

**Default**: false

---

### kvstore_flush_interval

Flush interval of the key-value store.

**Units**: milliseconds

**Default**: 10

---

### kvstore_max_segment_size

Maximum size of a segment of the key-value store.

**Type**: unsigned integer

**Units**: bytes

**Default**: 16777216 (16 MiB)

---

### log_segment_size

> Default size of log segment. Per topic, it is superseded by topic property `segment.bytes`.

Reviewer 2: If `segment.bytes` is a topic property, why isn't it included in the topic properties section?

Author: Rather than being a tunable cluster property for topics, `segment.bytes` is one of the topic properties that are not yet documented in a reference. That'll be done for issue #958 .

**Type**: unsigned 64-bit integer

**Units**: bytes

**Default**: 1073741824 (1 GiB)

**Range**: [1 MiB, ...]

---

### log_segment_size_max

Upper limit of topic `segment.bytes`. Higher values are clamped to this limit. 

If `null`, no limit is applied.

**Type**: unsigned 64-bit integer

**Units**: bytes

**Default**: null (no limit applied)

---

### log_segment_size_min

Lower limit of topic `segment.bytes`. Lower values are clamped to this limit. 

If `null`, no limit is applied.

**Type**: unsigned 64-bit integer

**Units**: bytes

**Default**: 1048576 (1 MiB)

---

### log_segment_size_jitter_percent

> Configures the amount of size variation added to a segment used for each partition: a percentage value in the range [0, log_segment_size_jitter_percent] is randomly chosen and that percentage of the original segment size limit is added. 

Reviewer 2: ```suggestion
Configures the amount of size variation added to a segment used for each partition: a percentage value in the range [0, `log_segment_size_jitter_percent`] is randomly chosen and that percentage of the original segment size limit is added. 
```

Applying segment size jitter is a good defensive measure to space out the overhead created from segment rolls over time.

**Type**: unsigned short

**Units**: percent of segment size that

**Default**: 5

**Range**: [0, 99]

---

### max_compacted_log_segment_size

Max compacted segment size after consolidation.

**Type**: unsigned integer

**Units**: bytes

**Default**: 5368709120 (5 GiB)

---

### readers_cache_eviction_timeout_ms

Duration after which inactive readers are evicted from the cache of readers.

**Units**: milliseconds

**Default**: 30000 (30 seconds)

---

### reclaim_batch_cache_min_free

Minimum amount of free memory maintained by the batch cache.

**Type**: unsigned integer

**Units**: bytes

**Default**: 67108864 (64 MiB)

---

### reclaim_growth_window

> Starting from the last point in time where memory was reclaimed from the batch cache, this is the duration during which the amount of memory to reclaim grows at a significant amount, based on heuristics about the amount of available memory.  

Reviewer 2: ```suggestion
Starting from the last point in time when memory was reclaimed from the batch cache, this is the duration during which the amount of memory to reclaim grows at a significant rate, based on heuristics about the amount of available memory.  
```

**Units**: milliseconds

**Default**: 3000

**Range**: <= [reclaim_stable_window](#reclaim_stable_window)

---

### reclaim_max_size

Maximum batch cache reclaim size.

**Units**: bytes

**Default**: 4194304 (4 MiB)

---

### reclaim_min_size

Minimum batch cache reclaim size.

**Units**: bytes

**Default**: 131072 (128 KiB)

---

### reclaim_stable_window

> If the duration since the last time memory was reclaimed has been longer than this property, the memory usage of the batch cache is considered as stable, so only the minimum size ([reclaim_min_size](#reclaim_min_size)) is set to be reclaimed.  

Reviewer 2: ```suggestion
If the duration since the last time memory was reclaimed has been longer than this property, the memory usage of the batch cache is considered stable, so only the minimum size ([reclaim_min_size](#reclaim_min_size)) is set to be reclaimed.  
```

**Units**: milliseconds

**Default**: 10000 (10 seconds)

**Range**: >= [reclaim_growth_window](#reclaim_growth_window)

---

### release_cache_on_segment_roll

Flag for whether to release cache when a full segment is rolled.

**Type**: boolean

**Default**: false

---

### replicate_append_timeout_ms

> Timeout for append entries RPCs issued while replicating entries.

Reviewer 2: ```suggestion
Timeout for append entry's RPCs issued while replicating entries.
```
Or 
Timeout for append entry RPCs issued while replicating entries.

**Units**: milliseconds

**Default**: 3000

---

### segment_appender_flush_timeout_ms

Maximum duration until buffered data is written.

**Units**: milliseconds

**Default**: 1000

---

### segment_fallocation_step

> Size of a file allocation (fallocation) step of a segment.

Reviewer 2: ```suggestion
Size of a file allocation (_f_ allocation) step of a segment.
```
